### PR TITLE
fix: remove `R_SEXP_to_hrg` use `R_SEXP_to_hrg_copy` to avoid memory leak

### PR DESCRIPTION
**Pull request by Antonov548**
Closes https://github.com/igraph/rigraph/issues/1073

### CODE REVIEW

File: src/rinterface.c [hunk 1]
```diff
@@ -7982,11 +7982,17 @@ SEXP R_igraph_hrg_sample(SEXP hrg) {
 
   SEXP r_result;
                                         /* Convert input */
-  R_SEXP_to_hrg(hrg, &c_hrg);
+  if (0 != R_SEXP_to_hrg_copy(hrg, &c_hrg)) {
+    igraph_error("", __FILE__, __LINE__, IGRAPH_ENOMEM);
```
Comment (krlmlr) [inline src/rinterface.c:7986]: In the Stimulus template:

```suggestion
    igraph_error("internal error in R_SEXP_to_hrg_copy()", __FILE__, __LINE__, IGRAPH_ENOMEM);
```

File: src/rinterface.c [hunk 2]
```diff
@@ -7982,11 +7982,17 @@ SEXP R_igraph_hrg_sample(SEXP hrg) {
...
+  }
+  IGRAPH_FINALLY(igraph_hrg_destroy, &c_hrg);
                                         /* Call igraph */
   IGRAPH_R_CHECK(igraph_hrg_sample(&c_hrg, &c_sample));
 
                                         /* Convert output */
+  PROTECT(hrg=R_igraph_hrg_to_SEXP(&c_hrg));
```
Comment (krlmlr) [inline src/rinterface.c:7993]: This adds a `PROTECT()` without an `UNPROTECT()` . 

- Which logic is typically responsible for adding/counting `UNPROTECT()` calls?
- The converted `hrg` is never used. @szhorvat: is the `INOUT` signature for this argument in the `functions.yaml` correct?

Comment (krlmlr) [reply src/rinterface.c:7993]: This must be resolved before we can merge this.

File: src/rinterface.c [hunk 3]
```diff
@@ -7951,7 +7951,7 @@ SEXP R_igraph_hrg_fit(SEXP graph, SEXP hrg, SEXP start, SEXP steps) {
                                         /* Convert input */
   R_SEXP_to_igraph(graph, &c_graph);
   if (0 != R_SEXP_to_hrg_copy(hrg, &c_hrg)) {
-    igraph_error("", __FILE__, __LINE__, IGRAPH_ENOMEM);
+    igraph_error("internal error in R_SEXP_to_hrg_copy()", __FILE__, __LINE__, IGRAPH_ENOMEM);
```
Comment (szhorvat) [inline src/rinterface.c:7954]:
```suggestion
    igraph_error("Insufficient memory to create HRG object.", __FILE__, __LINE__, IGRAPH_ENOMEM);
```

Please don't use the phrasing "internal error" for errors that are expected, such as out-of-memory.

File: tools/stimulus/types-RC.yaml [hunk 4]
```diff
@@ -58,6 +58,10 @@ INT:
         IN: |-
             IGRAPH_R_CHECK_INT(%I%);
             %C% = INTEGER(%I%)[0];
+    OUTCONV:
+        OUT: |-
+            PROTECT(%I%=NEW_NUMERIC(1));
+            REAL(%I%)[0]=(double) %C%;
```
Comment (krlmlr) [inline tools/stimulus/types-RC.yaml:64]: It's fine to work with `INTEGER()` here too, as the returned numbers will be very small.

In the future, can you please open a separate PR in such cases?

File: tools/stimulus/types-RC.yaml [hunk 5]
```diff
@@ -706,18 +710,26 @@ EDGE_ATTRIBUTE_COMBINATION:
...
-        IN: R_SEXP_to_hrg(%I%, &%C%);
+        IN: |-
+            if (0 != R_SEXP_to_hrg_copy(%I%, &%C%)) {
+              igraph_error("internal error in R_SEXP_to_hrg_copy()", __FILE__, __LINE__, IGRAPH_ENOMEM);
+            }
+            IGRAPH_FINALLY(igraph_hrg_destroy, &%C%);
         INOUT: |-
             if (0 != R_SEXP_to_hrg_copy(%I%, &%C%)) {
-              igraph_error("", __FILE__, __LINE__, IGRAPH_ENOMEM);
+              igraph_error("internal error in R_SEXP_to_hrg_copy()", __FILE__, __LINE__, IGRAPH_ENOMEM);
```
Comment (szhorvat) [inline tools/stimulus/types-RC.yaml:720]: this one needs a fix too and then regenerate

Comment (krlmlr) [reply tools/stimulus/types-RC.yaml:720]: @Antonov548: Can you please tweak according to Szabolcs's comments?

**Comment by maelle** [comment]
@Antonov548 friendly reminder of https://github.com/igraph/rigraph/pull/1075#discussion_r1449278866 :smile_cat:

File: tools/stimulus/types-RC.yaml [hunk 6]
```diff
@@ -76,6 +76,21 @@ CSTRING:
         OUT: |-
             PROTECT(%I% = Rf_mkCharLenCE(%C%, strlen(%C%), CE_UTF8));
 
+DOUBLE:
```
Comment (szhorvat) [inline tools/stimulus/types-RC.yaml:79]: What is the purpose of this type and where it is used in `functions.yaml`? It should always be the same as `REAL`.

**Comment by szhorvat** [comment]
@krlmlr @Antonov548 This PR has started to diverge form its original purpose. Would it be better to merge it and add unrelated fixes to independent PRs?

**Comment by Antonov548** [comment]
> @krlmlr @Antonov548 This PR has started to diverge form its original purpose. Would it be better to merge it and add unrelated fixes to independent PRs?

As soon as you happy with changes. I also think that this PR already following not the main goal.

**Comment by szhorvat** [comment]
I'm happy.

File: tools/stimulus/types-RC.yaml (hunk 6 above)
Comment (ntamas) [reply tools/stimulus/types-RC.yaml:79]: Nope :(

**Pull request merged**